### DISCUSSION TITLE
how to reference callback that defined under the components to the path operation?

### DISCUSSION THREAD

**Discussion by RSH17** [OAI/OpenAPI-Specification#3373] [Q&A] [answered]
when the callbacks are defined under the components and need to be referenced using **$ref**. how do define the $ref?
is this mentioned method okay to define the **$ref** under the callbacks? But the Swagger Editor gives an error.

{
  "openapi": "3.1.0",
  "info": {
    "title": "My API",
    "version": "1.0.0"
  },
  "components": {
    "callbacks": {
      "myCallback": {
        "{$request.path}/success": {
          "post": {
            "requestBody": {
              "content": {
                "application/json": {
                  "schema": {
                    "type": "object"
                  }
                }
              }
            },
            "responses": {
              "200": {
                "description": "Callback was successful"
              },
              "400": {
                "description": "Bad request"
              }
            }
          }
        },
        "{$request.path}/failure": {
          "post": {
            "requestBody": {
              "content": {
                "application/json": {
                  "schema": {
                    "type": "object"
                  }
                }
              }
            },
            "responses": {
              "500": {
                "description": "Internal server error"
              }
            }
          }
        }
      }
    }
  },
  "paths": {
    "/example": {
      "post": {
        "requestBody": {
          "required": true,
          "content": {
            "application/json": {
              "schema": {
                "type": "object"
              }
            }
          }
        },
        "callbacks": {
          "$ref": "#/components/callbacks/myCallback"
        },
        "responses": {
          "200": {
            "description": "Success response"
          },
          "400": {
            "description": "Bad request"
          }
        }
      }
    }
  }
}

![test](https://github.com/OAI/OpenAPI-Specification/assets/56576410/ce4352bb-d0ba-4e0c-98a1-c52772ba2b8f)

**Comment by hkosova** [comment, marked as answer]
Replace
```
"callbacks": {
"$ref": "#/components/callbacks/myCallback"
},
```
with
```
"callbacks": {
  "myCallback": {
    "$ref": "#/components/callbacks/myCallback"
  }
},
```

**Reply by eric-gonzalez-tfs** [reply]
@hkosova - This answer did not work for me.

![image](https://github.com/user-attachments/assets/a7d82d5a-4651-4c7f-9d3f-39617ebc3a71)


```json
{
   "openapi":"3.1.0",
   "info":{
      "title":"My API",
      "version":"1.0.0"
   },
   "components":{
      "callbacks":{
         "myCallback":{
            "{$request.path}/success":{
               "post":{
                  "requestBody":{
                     "content":{
                        "application/json":{
                           "schema":{
                              "type":"object"
                           }
                        }
                     }
                  },
                  "responses":{
                     "200":{
                        "description":"Callback was successful"
                     },
                     "400":{
                        "description":"Bad request"
                     }
                  }
               }
            },
            "{$request.path}/failure":{
               "post":{
                  "requestBody":{
                     "content":{
                        "application/json":{
                           "schema":{
                              "type":"object"
                           }
                        }
                     }
                  },
                  "responses":{
                     "500":{
                        "description":"Internal server error"
                     }
                  }
               }
            }
         }
      }
   },
   "paths":{
      "/example":{
         "post":{
            "requestBody":{
               "required":true,
               "content":{
                  "application/json":{
                     "schema":{
                        "type":"object"
                     }
                  }
               }
            },
            "callbacks":{
               "myCallback":{
                  "$ref":"#/components/callbacks/myCallback"
               }
            },
            "responses":{
               "200":{
                  "description":"Success response"
               },
               "400":{
                  "description":"Bad request"
               }
            }
         }
      }
   }
}
```

**Reply by hkosova** [reply]
@eric-gonzalez-tfs your callback syntax is correct. The error reported by the tool is a false positive. Open an issue with the tool vendor.

If the tool on your image is Swagger Editor - it's a known issue https://github.com/swagger-api/swagger-editor/issues/4682

**Reply by eric-gonzalez-tfs** [reply]
Thanks for the response @hkosova - I indeed was using the Swagger Editor. I appreciate the link to the open issue!

**Comment by handrews** [comment]
Closing as resolved.
